### PR TITLE
Fixed Homunculus Lif skills

### PR DESCRIPTION
**Pull request by Daegaladh**
* **Addressed Issue(s)**: N/A

* **Server Mode**: Both

* **Description of Pull Request**:
  - Fixes status flag **RemoveOnMapWarp** not removing **homunculus** status
  - Fixes homunculus speed not properly recalculating when hom_setting HOMSET_COPY_SPEED is disabled
  - Fixes homunculus status not clearing on vaporize
  - Adds RemoveFromHomOnWarp and RemoveFromHomOnMapWarp status flags
  - Fixes Lif skills behavior:
      -  **HLIF_HEAL (Healing Hands)**:
         - **Renewal**: Not ignoring GTB card
         - Removed from the undead damage check as as it works with undead armor or change undead status on official servers (previously didn't work anyway as the other healing skills checks for player's sd)
      - **HLIF_AVOID (Urgent Escape)**:
         - Speed rates for **homunculus** (40 * level) and **master** (10 * level)
         - Speed calculations not adding **homunculus** speed buff/debuff on top of the current speed
         - **Renewal**: **Homunculus** speed not re-synchronizing with their **master**'s (and reapply their own buff/debuff on top)
         - Added missing flags for **SC_AVOID** status:
           - **Pre-renewal**: Not getting removed from the **homunculus** on **Teleport**, **Map change** and **Vaporize**
           - **Renewal**: Not getting removed from the **homunculus** on **Map change** and **Vaporize**
           - **Both**: Not getting removed from both the **homunculus** and the **master** on **Relog**
      - **HLIF_CHANGE (Mental Charge)**:
         - **Pre-Renewal**: Not restoring full hp and sp on use
         - Added / fixed flags for **SC_CHANGE** status:
           - **Renewal**: Getting removed from the **homunculus** on **Teleport**
           - **Both**: Not getting removed from both the **homunculus** and the **master** on **Relog**

Official behavior:
On renewal the homunculus speed is calculated by copying the master's speed, then applying the homunculus speed buffs/debuffs on top. Also, speed is recalculated each time the master receives a bonus.
On pre-renewal, on the other hand, the speed is copied from master on each refresh (teleport, change maps, relog and vaporize/call). The speed is not recalculated when you receive a buff/debuff, but gets copied on refresh after receiving it (e.g. if the master receives Increase AGI, then teleports, the new speed is copied by the homunculus).

Homunculus statuses always get cleared on vaporize on both modes.

HLIF_HEAL is blocked by GTB card on renewal, but not on pre-renewal. Slso doesn't damage the player if Evil Druid card is equipped on both modes.

HLIF_AVOID has different speed rates for the **homunculus** (40 * level) and **master** (10 * level).
SC_AVOID
  On pre-re it's removed from the homunculus on teleport and map change but not from the master, it gets removed from both on relog
  On renewal, it's removed from the homunculus on map change (not teleport) but not from the master, it gets removed from both on relog

HLIF_CHANGE fully restores the homunculus' HP and SP on pre-re. On both modes the homunculus HP and SP is set to 10 when the effect naturally expires (not if canceled by changing maps, relog, etc).
  On pre-re it's removed on teleport and map change and relog.
  On renewal, it's removed on map change and relog.

### CODE REVIEW

File: src/map/pc.cpp [hunk 1]
```diff
@@ -7008,6 +7008,19 @@ enum e_setpos pc_setpos(map_session_data* sd, unsigned short mapindex, int x, in
 
 	if( hom_is_active(sd->hd) )
 	{
+		status_change* hom_sc = status_get_sc(&sd->hd->bl);
+
+		if ( hom_sc != nullptr && hom_sc->count ) {
+			for (const auto &it : status_db) {
+				if (hom_sc->getSCE(it.first)) {
+					if (it.second->flag[SCF_REMOVEFROMHOMONWARP]) {
+						ShowDebug("pc_setpos: Removing status %d from homunculus %d on map change\n", it.first, sd->hd->bl.id);
```
Comment (Lemongrass3110) [inline src/map/pc.cpp:7017]:
```suggestion
```
Leftover from testing?

File: src/map/pc.cpp [hunk 2]
```diff
@@ -7008,6 +7008,19 @@ enum e_setpos pc_setpos(map_session_data* sd, unsigned short mapindex, int x, in
 
 	if( hom_is_active(sd->hd) )
 	{
+		status_change* hom_sc = status_get_sc(&sd->hd->bl);
+
+		if ( hom_sc != nullptr && hom_sc->count ) {
```
Comment (Lemongrass3110) [inline src/map/pc.cpp:7013]:
```suggestion
		if ( hom_sc != nullptr && hom_sc->count > 0 ) {
```

File: src/map/status.cpp [hunk 3]
```diff
@@ -5841,13 +5841,28 @@ void status_calc_bl_main(struct block_list *bl, std::bitset<SCB_MAX> flag)
 	if(flag[SCB_SPEED]) {
 		status->speed = status_calc_speed(bl, sc, b_status->speed);
 
-		if( bl->type&BL_PC && !(sd && sd->state.permanent_speed) && status->speed < battle_config.max_walk_speed )
-			status->speed = battle_config.max_walk_speed;
+		if (bl->type & BL_PC && sd != nullptr) {
```
Comment (Lemongrass3110) [inline src/map/status.cpp:5844]:
```suggestion
		if (sd != nullptr) {
```
sd was already assigned with BL_CAST and checked the bl->type

File: src/map/status.cpp [hunk 4]
```diff
@@ -5841,13 +5841,28 @@ void status_calc_bl_main(struct block_list *bl, std::bitset<SCB_MAX> flag)
...
+			if (!sd->state.permanent_speed && status->speed < battle_config.max_walk_speed)
+				status->speed = battle_config.max_walk_speed;
 
+			// Recalculate homunculus speed if the player receives a speed buff/debuff
+			if (hom_is_active(sd->hd))
+				status_calc_bl(&sd->hd->bl, { SCB_SPEED, SCB_HOM });
+		}
 		if( bl->type&BL_PET && ((TBL_PET*)bl)->master)
```
Comment (Lemongrass3110) [inline src/map/status.cpp:5852]:
```suggestion
		if( bl->type == BL_PET && reinterpret_cast<pet_data*>( bl )->master)
```

File: src/map/status.cpp [hunk 5]
```diff
@@ -5841,13 +5841,28 @@ void status_calc_bl_main(struct block_list *bl, std::bitset<SCB_MAX> flag)
...
 			status->speed = status_get_speed(&((TBL_PET*)bl)->master->bl);
-		if( bl->type&BL_HOM && battle_config.hom_setting&HOMSET_COPY_SPEED && ((TBL_HOM*)bl)->master)
-			status->speed = status_get_speed(&((TBL_HOM*)bl)->master->bl);
+		if( bl->type & BL_HOM) {
```
Comment (Lemongrass3110) [inline src/map/status.cpp:5854]:
```suggestion
		if( bl->type == BL_HOM) {
```

File: src/map/status.cpp [hunk 6]
```diff
@@ -5841,13 +5841,28 @@ void status_calc_bl_main(struct block_list *bl, std::bitset<SCB_MAX> flag)
...
+			homun_data* hd = reinterpret_cast<homun_data*>(bl);
+
+			if (hd != nullptr && hd->master != nullptr) {
```
Comment (Lemongrass3110) [inline src/map/status.cpp:5857]:
```suggestion
			if (hd->master != nullptr) {
```
Cannot be null here.

File: src/map/status.cpp [hunk 7]
```diff
@@ -11444,7 +11459,11 @@ int status_change_start(struct block_list* src, struct block_list* bl,enum sc_ty
 			// val4 signals autoprovoke.
 			break;
 		case SC_AVOID:
-			// val2 = 10*val1; // Speed change rate.
+			// Speed change rate.
+			if (bl->type & BL_HOM)
```
Comment (Lemongrass3110) [inline src/map/status.cpp:11463]:
```suggestion
			if (bl->type == BL_HOM)
```

File: src/map/status.hpp [hunk 8]
```diff
@@ -2947,6 +2947,7 @@ enum e_scb_flag : uint8 {
 
 	// Extra Flags
 	SCB_DYE, // force cloth-dye change to 0 to avoid client crashes.
+	SCB_HOM, // Force recalculate homunculus speed
```
Comment (Lemongrass3110) [inline src/map/status.hpp:2950]: There is no check for it?
What do you want to do with it?

File: src/map/status.cpp [hunk 9]
```diff
@@ -11444,7 +11459,11 @@ int status_change_start(struct block_list* src, struct block_list* bl,enum sc_ty
 			// val4 signals autoprovoke.
 			break;
 		case SC_AVOID:
```
Comment (Lemongrass3110) [inline src/map/status.cpp:11461]: If possible update SC_AVOID in /doc/status_change.txt the whole description is missing anyway.

File: src/map/homunculus.cpp [hunk 10]
```diff
@@ -285,6 +285,16 @@ int hom_vaporize(map_session_data *sd, int flag)
 	if (flag == HOM_ST_REST && get_percentage(hd->battle_status.hp, hd->battle_status.max_hp) < 80)
 		return 0;
 
+	status_change* sc = status_get_sc(&hd->bl);
+
+	if (sc != nullptr && sc->count) {
```
Comment (Lemongrass3110) [inline src/map/homunculus.cpp:290]:
```suggestion
	if (sc != nullptr && sc->count > 0) {
```

File: src/map/homunculus.cpp [hunk 11]
```diff
@@ -285,6 +285,16 @@ int hom_vaporize(map_session_data *sd, int flag)
...
+		for (const auto& it : status_db) {
+			if (sc->getSCE(it.first))
+				if (it.second->flag[SCF_REMOVEFROMHOMONVAPORIZE])
+					status_change_end(&hd->bl, static_cast<sc_type>(it.first));
+		}
```
Comment (Lemongrass3110) [inline src/map/homunculus.cpp:295]:
```suggestion
		status_db.removeByStatusFlag( &hd->bl, { SCF_REMOVEFROMHOMONVAPORIZE } );
```

File: src/map/pc.cpp [hunk 12]
```diff
@@ -7008,6 +7008,19 @@ enum e_setpos pc_setpos(map_session_data* sd, unsigned short mapindex, int x, in
...
+			for (const auto &it : status_db) {
+				if (hom_sc->getSCE(it.first)) {
+					if (it.second->flag[SCF_REMOVEFROMHOMONWARP]) {
+						ShowDebug("pc_setpos: Removing status %d from homunculus %d on map change\n", it.first, sd->hd->bl.id);
+						status_change_end(&sd->hd->bl, static_cast<sc_type>(it.first));
+					}
+				}
+			}
```
Comment (Lemongrass3110) [inline src/map/pc.cpp:7021]:
```suggestion
			status_db.removeByStatusFlag( &sd->hd->bl, { SCF_REMOVEFROMHOMONWARP } );
```

File: src/map/status.cpp [hunk 13]
```diff
@@ -5841,17 +5841,57 @@ void status_calc_bl_main(struct block_list *bl, std::bitset<SCB_MAX> flag)
...
+					sd->hd->battle_status.speed = status_calc_speed(&sd->hd->bl, &sd->hd->sc, sd->hd->battle_status.speed);
+				}
+#endif
+			}
+			break;
+
+		case BL_PET:{
+			pet_data* pd = reinterpret_cast<pet_data*>(bl);
+
+			if (pd->master)
```
Comment (aleos89) [inline src/map/status.cpp:5865]:
```suggestion
			if (pd != nullptr && pd->master != nullptr)
```

File: src/map/status.cpp [hunk 14]
```diff
@@ -5841,17 +5841,57 @@ void status_calc_bl_main(struct block_list *bl, std::bitset<SCB_MAX> flag)
...
-		if( bl->type&BL_HOM && battle_config.hom_setting&HOMSET_COPY_SPEED && ((TBL_HOM*)bl)->master)
-			status->speed = status_get_speed(&((TBL_HOM*)bl)->master->bl);
-		if( bl->type&BL_MER && ((TBL_MER*)bl)->master)
-			status->speed = status_get_speed(&((TBL_MER*)bl)->master->bl);
-		if( bl->type&BL_ELEM && ((TBL_ELEM*)bl)->master)
-			status->speed = status_get_speed(&((TBL_ELEM*)bl)->master->bl);
+		case BL_HOM:{
+			homun_data* hd = reinterpret_cast<homun_data*>(bl);
+
+			if (hd->master) {
```
Comment (aleos89) [inline src/map/status.cpp:5872]:
```suggestion
			if (hd != nullptr && hd->master != nullptr) {
```

File: src/map/status.cpp [hunk 15]
```diff
@@ -5841,17 +5841,57 @@ void status_calc_bl_main(struct block_list *bl, std::bitset<SCB_MAX> flag)
...
+
+				// Homunculus speed buff/debuffs applies over the current speed
+				status->speed = status_calc_speed(bl, &hd->sc, status->speed);
+			}
+		} break;
+
+		case BL_MER:{
+			s_mercenary_data* mc = reinterpret_cast<s_mercenary_data*>(bl);
+
+			if (mc->master)
```
Comment (aleos89) [inline src/map/status.cpp:5884]:
```suggestion
			if (mc != nullptr && mc->master != nullptr)
```

File: src/map/status.cpp [hunk 16]
```diff
@@ -5841,17 +5841,57 @@ void status_calc_bl_main(struct block_list *bl, std::bitset<SCB_MAX> flag)
...
+				status->speed = status_get_speed(&mc->master->bl);
+		} break;
+
+		case BL_ELEM:{
+			s_elemental_data* ed = reinterpret_cast<s_elemental_data*>(bl);
+
+			if (ed->master)
```
Comment (aleos89) [inline src/map/status.cpp:5891]:
```suggestion
			if (ed != nullptr && ed->master != nullptr)
```

**Review by aleos89 — APPROVED** [review]
Overall looks good! This is from a mobile review but just a few safety checks is all i could find for now!

File: src/map/skill.cpp [hunk 17]
```diff
@@ -10717,11 +10717,15 @@ int skill_castend_nodamage_id (struct block_list *src, struct block_list *bl, ui
 			status_heal(bl, i, 0, 0);
 		}
 		break;
-	//Homun single-target support skills [orn]
+	// Homun single-target support skills [orn]
+	case HLIF_CHANGE:
+#ifndef RENEWAL
+		status_heal(bl, status_get_max_hp(bl), status_get_max_sp(bl), 0);
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:10723]:
```suggestion
		status_percent_heal(bl, 100, 100);
```

File: src/map/status.cpp [hunk 18]
```diff
@@ -5841,17 +5841,57 @@ void status_calc_bl_main(struct block_list *bl, std::bitset<SCB_MAX> flag)
...
-		if( bl->type&BL_HOM && battle_config.hom_setting&HOMSET_COPY_SPEED && ((TBL_HOM*)bl)->master)
-			status->speed = status_get_speed(&((TBL_HOM*)bl)->master->bl);
-		if( bl->type&BL_MER && ((TBL_MER*)bl)->master)
-			status->speed = status_get_speed(&((TBL_MER*)bl)->master->bl);
-		if( bl->type&BL_ELEM && ((TBL_ELEM*)bl)->master)
-			status->speed = status_get_speed(&((TBL_ELEM*)bl)->master->bl);
+		case BL_HOM:{
+			homun_data* hd = reinterpret_cast<homun_data*>(bl);
+
+			if (hd != nullptr && hd->master != nullptr) {
```
Comment (Lemongrass3110) [inline src/map/status.cpp:5872]:
```suggestion
			if (hd->master != nullptr) {
```

File: src/map/status.cpp [hunk 19]
```diff
@@ -5841,17 +5841,57 @@ void status_calc_bl_main(struct block_list *bl, std::bitset<SCB_MAX> flag)
...
+
+				// Homunculus speed buff/debuffs applies over the current speed
+				status->speed = status_calc_speed(bl, &hd->sc, status->speed);
+			}
+		} break;
+
+		case BL_MER:{
+			s_mercenary_data* mc = reinterpret_cast<s_mercenary_data*>(bl);
+
+			if (mc != nullptr && mc->master != nullptr)
```
Comment (Lemongrass3110) [inline src/map/status.cpp:5884]:
```suggestion
			if (mc->master != nullptr)
```

File: src/map/status.cpp [hunk 20]
```diff
@@ -5841,17 +5841,57 @@ void status_calc_bl_main(struct block_list *bl, std::bitset<SCB_MAX> flag)
...
+				status->speed = status_get_speed(&mc->master->bl);
+		} break;
+
+		case BL_ELEM:{
+			s_elemental_data* ed = reinterpret_cast<s_elemental_data*>(bl);
+
+			if (ed != nullptr && ed->master != nullptr)
```
Comment (Lemongrass3110) [inline src/map/status.cpp:5891]:
```suggestion
			if (ed->master != nullptr)
```

File: src/map/status.cpp [hunk 21]
```diff
@@ -5841,17 +5841,57 @@ void status_calc_bl_main(struct block_list *bl, std::bitset<SCB_MAX> flag)
...
+					sd->hd->battle_status.speed = status_calc_speed(&sd->hd->bl, &sd->hd->sc, sd->hd->battle_status.speed);
+				}
+#endif
+			}
+			break;
+
+		case BL_PET:{
+			pet_data* pd = reinterpret_cast<pet_data*>(bl);
+
+			if (pd != nullptr && pd->master != nullptr)
```
Comment (Lemongrass3110) [inline src/map/status.cpp:5865]:
```suggestion
			if (pd->master != nullptr)
```

File: src/map/status.cpp [hunk 22]
```diff
@@ -5841,17 +5841,57 @@ void status_calc_bl_main(struct block_list *bl, std::bitset<SCB_MAX> flag)
 	if(flag[SCB_SPEED]) {
 		status->speed = status_calc_speed(bl, sc, b_status->speed);
 
-		if( bl->type&BL_PC && !(sd && sd->state.permanent_speed) && status->speed < battle_config.max_walk_speed )
-			status->speed = battle_config.max_walk_speed;
+		switch (bl->type) {
+		case BL_PC:
+			if (sd != nullptr) {
```
Comment (Lemongrass3110) [inline src/map/status.cpp:5846]:
```suggestion
```
Everything has to be moved out one tab then.

File: src/map/status.cpp [hunk 23]
```diff
@@ -5841,17 +5841,57 @@ void status_calc_bl_main(struct block_list *bl, std::bitset<SCB_MAX> flag)
...
+				// Recalculate homunculus speed if the player receives a speed buff/debuff
+				if (hom_is_active(sd->hd)) {
+					if (battle_config.hom_setting & HOMSET_COPY_SPEED)
+						sd->hd->battle_status.speed = status_get_speed(&sd->bl);
+
+					// Homunculus speed buff/debuffs applies over the current speed
+					sd->hd->battle_status.speed = status_calc_speed(&sd->hd->bl, &sd->hd->sc, sd->hd->battle_status.speed);
+				}
+#endif
+			}
```
Comment (Lemongrass3110) [inline src/map/status.cpp:5859]:
```suggestion
```

**Review by Lemongrass3110 — APPROVED** [review]

**Pull request merged**